### PR TITLE
Get rid of GPIO event source and its corresponding thread

### PR DESCRIPTION
**Pull request by erikcorry**

### CODE REVIEW

**Comment by erikcorry** [comment]
This makes the EventQueueEventSource less generic, adding GPIO code to it.  Perhaps there's some way to make it neater.

The Resources for GPIO are a bit different from the Resoures that EventQueueEventSource has handled until now.  They don't have a FreeRTOS queue in them - they all share one queue.  When an event arrives on that queue (from the interrupt) we check all the resources to see if one of them has the right pin number.  Before we do that we already got the level of the correct pin in order to sample it as early as possible.

File: src/event_sources/ev_queue_esp32.cc [hunk 1]
```diff
@@ -78,30 +100,36 @@ void EventQueueEventSource::entry() {
 
 void EventQueueEventSource::on_register_resource(Locker& locker, Resource* r) {
   auto evq_res = static_cast<EventQueueResource*>(r);
```
Comment (kasperl) [inline src/event_sources/ev_queue_esp32.cc:102]: I would rename this to resource.

File: src/event_sources/ev_queue_esp32.cc [hunk 2]
```diff
@@ -78,30 +100,36 @@ void EventQueueEventSource::entry() {
...
+      word data;
+      while (evq_res->receive_event(&data)) {
+        dispatch(locker, r, data);
+      }
+    } while (xQueueAddToSet(queue, _queue_set) != pdPASS);
+  }
 }
 
 void EventQueueEventSource::on_unregister_resource(Locker& locker, Resource* r) {
   auto evq_res = static_cast<EventQueueResource*>(r);
```
Comment (kasperl) [inline src/event_sources/ev_queue_esp32.cc:119]: resource

File: src/resources/gpio_esp32.cc [hunk 3]
```diff
@@ -49,14 +49,45 @@ ResourcePool<int, -1> gpio_pins(
     32, 33, 34, 35, 36, 37, 38, 39
 );
 
+class GPIOResource : public EventQueueResource {
+public:
```
Comment (kasperl) [inline src/resources/gpio_esp32.cc:53]: One space indent.

File: src/resources/gpio_esp32.cc [hunk 4]
```diff
@@ -49,14 +49,45 @@ ResourcePool<int, -1> gpio_pins(
...
+  GPIOResource(ResourceGroup* group, int pin)
+      // GPIO resources share a queue, which is always on the event source, so pass null.
+      : EventQueueResource(group, null)
+      , _pin(pin) {}
+
+  int pin() const { return _pin; }
+
+  bool check_gpio(word pin) override;
+
+private:
```
Comment (kasperl) [inline src/resources/gpio_esp32.cc:65]: One space indent.

File: src/event_sources/ev_queue_esp32.cc [hunk 5]
```diff
@@ -78,30 +100,36 @@ void EventQueueEventSource::entry() {
...
-  int attempts = 0;
-  do {
-    if (attempts++ > 16) FATAL("couldn't register event resource");
-    word data;
-    while (evq_res->receive_event(&data)) {
-      dispatch(locker, r, data);
-    }
-  } while (xQueueAddToSet(evq_res->queue(), _queue_set) != pdPASS);
+  QueueHandle_t queue = evq_res->queue();
+  if (queue) {
```
Comment (kasperl) [inline src/event_sources/ev_queue_esp32.cc:104]: Early bailout by returning on !queue?

File: src/event_sources/ev_queue_esp32.cc [hunk 6]
```diff
@@ -78,30 +100,36 @@ void EventQueueEventSource::entry() {
...
-  int attempts = 0;
-  do {
-    if (attempts++ > 16) FATAL("couldn't unregister event resource");
-    word data;
-    while (evq_res->receive_event(&data)) {
-      // Don't dispatch while unregistering.
-    }
-  } while (xQueueRemoveFromSet(evq_res->queue(), _queue_set) != pdPASS);
+  QueueHandle_t queue = evq_res->queue();
+  if (queue) {
```
Comment (kasperl) [inline src/event_sources/ev_queue_esp32.cc:121]: Early bailout by returning on !queue?

File: src/resources/gpio_esp32.cc [hunk 7]
```diff
@@ -49,14 +49,45 @@ ResourcePool<int, -1> gpio_pins(
...
 class GPIOResourceGroup : public ResourceGroup {
  public:
   TAG(GPIOResourceGroup);
   explicit GPIOResourceGroup(Process* process)
-      : ResourceGroup(process, GPIOEventSource::instance()) {}
+      : ResourceGroup(process, EventQueueEventSource::instance()) {
+    queue = EventQueueEventSource::instance()->gpio_queue();
+  }
+
+  virtual void on_register_resource(Resource* r) {
```
Comment (kasperl) [inline src/resources/gpio_esp32.cc:77]: I would probably move the code for on_register and on_unregister out of the class definition.

File: src/event_sources/ev_queue_esp32.cc [hunk 8]
```diff
@@ -30,8 +33,14 @@ EventQueueEventSource::EventQueueEventSource()
     : EventSource("EVQ")
     , Thread("EVQ")
     , _stop(xSemaphoreCreateBinary())
+    , _gpio_queue(xQueueCreate(32, sizeof(word)))
     , _queue_set(xQueueCreateSet(32)) {
   xQueueAddToSet(_stop, _queue_set);
+  xQueueAddToSet(_gpio_queue, _queue_set);
+
+  SystemEventSource::instance()->run([&]() -> void {
+    FATAL_IF_NOT_ESP_OK(gpio_install_isr_service(ESP_INTR_FLAG_IRAM));
```
Comment (kasperl) [inline src/event_sources/ev_queue_esp32.cc:42]: Do we ever uninstall the ISR service with the new code?

File: src/event_sources/ev_queue_esp32.cc [hunk 9]
```diff
@@ -65,7 +75,19 @@ void EventQueueEventSource::entry() {
       return;
     }
 
-    // Then loop through all queues.
+    // See if there's a GPIO event.
+    word pin;
+    while (xQueueReceive(_gpio_queue, &pin, 0)) {
+      bool data = gpio_get_level(gpio_num_t(pin)) != 0;
+      for (auto r : resources()) {
+        auto evq_res = static_cast<EventQueueResource*>(r);
```
Comment (kasperl) [inline src/event_sources/ev_queue_esp32.cc:83]: resource

File: src/event_sources/ev_queue_esp32.cc [hunk 10]
```diff
@@ -65,7 +75,19 @@ void EventQueueEventSource::entry() {
...
+        if (evq_res->check_gpio(pin)) {
+          dispatch(locker, r, data);
+        }
+      }
+    }
+
+    // Then loop through other queues.
     for (auto r : resources()) {
       auto evq_res = static_cast<EventQueueResource*>(r);
```
Comment (kasperl) [inline src/event_sources/ev_queue_esp32.cc:92]: resource

**Review by kasperl — APPROVED** [review]
Looks good to me.

**Pull request merged**